uid/src/main/java/com/tomtom/kotlin/uid/Uid.kt

```diff
@@ -0,0 +1,217 @@
+            sb.append('0')
+        }
+        sb.append(lsb)
+        val value = sb.toString()
+        assert(value.length == 32)
+        return value
+    }
+
+    /**
+     * Method converts given String representation to [Uid] and compares it with this instance.
+     * A String value of "0-0-0-0-0" would match a [Uid] of "00000-0000-0000-000000000-00" or so.
+     *
+     * @param uid String representation od Uid.
+     * @return True in case String representation matches instance. False otherwise.
+     */
+    fun matchesFromString(uid: String) = this == fromString<T>(uid)
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Uid<*>) return false
+        if (uuid != other.uuid) return false
+        return true
+    }
+
+    override fun hashCode() = uuid.hashCode()
+
+    /**
+     * Returns string representation of this Uid. Opposite of [fromString].
+     *
+     * @return String representation of ID.
+     */
+    override fun toString() = uuid
+
+    companion object {
+        private const val UUID_DASH = '-'
+        private const val UUID_MIN_LENGTH = 9
+        private const val UUID_MAX_LENGTH = 36
+        private val UUID_DASH_POS = intArrayOf(8, 13, 18, 23)
+
+        /**
+         * Returns an ID if it is a valid UUID, or `null` if it's not.
+         *
+         * @param id String representation of UUID.
+         * @return Valid UUID, or null.
+         */
+        fun <T> fromStringIfValid(id: String?) =
+            if (id == null) {
+                null
+            } else try {
+                Uid<T>(id)
+            } catch (ignored: IllegalArgumentException) {
+                null
+            }
+
+        /**
+         * Instantiates a [Uid] with given ID as a string. Mainly used for deserialization.
+         * Opposite of [toString].
+         *
+         * @param <T> Uid type.
+         * @param id  String representation of ID.
+         * @return Uid.
+         * @throws IllegalArgumentException If name does not conform to the string
+         * representation as described in [UUID.toString]. Use [isValid] to make sure the
+         * string is valid.
+        </T> */
+        fun <T> fromString(id: String) = Uid<T>(id)
+
+        /**
+         * Instantiates a [Uid] with given ID as hex-formatted string. Opposite of [toHexString].
+         *
+         * @param <T> Uid type.
+         * @param id  Hex string representation of ID, must be exactly 32 characters long.
+         * @return Uid.
+         * @throws IllegalArgumentException If name does not conform to the string
+         * representation as described in [UUID.toString]. Use [isValid] to make sure the
+         * string is valid.
+        </T> */
+        fun <T> fromHexString(id: String): Uid<T> {
+            require(id.length == 32)
+            val msb: Long =
+                id.substring(0, 8).toLong(16) shl 32 or id.substring(8, 16).toLong(16)
```

Comment:
This is equivalent:

```
val msb = id.substring(0, 16).toLong(16)
val lsb = id.substring(16, 32).toLong(16)
```

Reply:
.

Reply:
Test fails with this. Not equivalent.

Reply:
Oh, strange, I tried it and it worked for me...